pkg/frontend/querymiddleware/astmapper/instant_splitting.go

```diff
@@ -426,30 +429,41 @@ func getRangeInterval(expr parser.Expr) time.Duration {
 	}
 }
 
-// getOffset returns the offset interval in the range vector expression
-// Returns 0 if no offset interval is found
-// Examples:
-//   * `count_over_time({app="foo"}[10m])` returns 0
-//   * `count_over_time({app="foo"}[10m]) offset 1m` returns 1m
-func getOffset(expr parser.Expr) time.Duration {
-	switch e := expr.(type) {
-	case *parser.AggregateExpr:
-		return getOffset(e.Expr)
-	case *parser.Call:
-		// Iterate over Call's arguments until a VectorSelector and a valid offset are found
-		for _, arg := range e.Args {
-			if argRangeInterval := getOffset(arg); argRangeInterval > 0 {
-				return argRangeInterval
-			}
-		}
-		return time.Duration(0)
-	case *parser.MatrixSelector:
-		return getOffset(e.VectorSelector)
-	case *parser.VectorSelector:
-		return e.OriginalOffset
-	default:
-		return 0
+// assertSplittableRangeInterval returns the range interval specified in the input expr and whether it is greater than
+// the configured split interval.
+func (i *instantSplitter) assertOffset(expr parser.Expr) (offset time.Duration, err error) {
+	offsets := getOffsets(expr)
+	if len(offsets) == 0 {
+		return time.Duration(0), nil
+	}
+	if len(offsets) > 1 {
+		return time.Duration(0), fmt.Errorf("found %d offsets while expecting at most 1", len(offsets))
 	}
+
+	// Possible to have negative offsets
+	offset = offsets[0]
+	return offset, nil
```

Comment:
[nit] Can be simplified.

```suggestion
	return offsets[0], nil
```